server/src/main/java/org/elasticsearch/script/AbstractSortScript.java

```diff
@@ -0,0 +1,106 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.script;
+
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.search.Scorable;
+import org.elasticsearch.ElasticsearchException;
+import org.elasticsearch.common.lucene.ScorerAware;
+import org.elasticsearch.index.fielddata.ScriptDocValues;
+import org.elasticsearch.search.lookup.LeafSearchLookup;
+import org.elasticsearch.search.lookup.SearchLookup;
+
+abstract class AbstractSortScript implements ScorerAware {
```

Comment:
I don't think this can be package private, as painless would not be able to access the methods from it?

Reply:
I'm also not sure there is really enough here to justify a super class. Can we just have the few lines duplicated and abstract this when we see more need?

Reply:
@rjernst I mainly made the `abstract` class because I figured there's quite a bit of magic with the way the lookups work and such and it wasn't so tricky to make this.

Also, it's fine for this to be package private. Painless uses `Class#getMethods()` which is documented as:

```java
     * Returns an array containing {@code Method} objects reflecting all the
     * public methods of the class or interface represented by this {@code
     * Class} object, including those declared by the class or interface and
     * those inherited from superclasses and superinterfaces.
```

Reply:
Thanks, that is good to know about the lookups, I didn't realize that. :)

I would at least rename this as I don't think it has to do with "sorting" at all, it is about making score accessible?

Reply:
@rjernst true, `AbstractScoreAwareScript` ? :)